### ISSUE TITLE
Resolver fails when preferred version of required feature is not available, even though a tolerated version is available

### ISSUE THREAD

**Issue by ericglau** [OpenLiberty/open-liberty#3470]
Some Open Liberty features have a required feature where the preferred version is only in WebSphere Liberty, but tolerates a version in Open Liberty.  For example, jsp-2.2 (in OL) requires servlet-3.0 (in WL) and tolerates 3.1 and 4.0 (in OL).

If we call the resolver with the set [jsp-2.2, servlet-3.1] in Open Liberty, since servlet-3.0 is not available, the resolver fails with an exception as below.  Instead, the resolver should handle the tolerates to allow resolving jsp-2.2 together with tolerated feature servlet-3.1.

```
com.ibm.ws.install.InstallException: CWWKF1385E: The [jsp-2.2, servlet-3.1] assets depend on com.ibm.websphere.appserver.servlet-3.0, which is not available in the IBM WebSphere Liberty Repository. If the required asset is a feature that you previously downloaded, install the missing feature by specifying the location of the feature ESA file on the featureManager install --location option, and try to install [jsp-2.2, servlet-3.1] again.
	at com.ibm.ws.install.internal.ExceptionUtils.create(ExceptionUtils.java:560)
	at com.ibm.ws.install.internal.ExceptionUtils.create(ExceptionUtils.java:133)
	at com.ibm.ws.install.internal.ExceptionUtils.create(ExceptionUtils.java:372)
	at com.ibm.ws.install.internal.InstallKernelMap.singleFileResolve(InstallKernelMap.java:531)
	at com.ibm.ws.install.internal.InstallKernelMap.get(InstallKernelMap.java:204)
	at com.ibm.ws.install.map.InstallMap.get(InstallMap.java:257)
	at net.wasdev.wlp.maven.plugins.server.InstallFeatureMojo.installFeatures(InstallFeatureMojo.java:393)
	at net.wasdev.wlp.maven.plugins.server.InstallFeatureMojo.doExecute(InstallFeatureMojo.java:198)
	at org.codehaus.mojo.pluginsupport.MojoSupport.execute(MojoSupport.java:122)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:134)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:208)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:154)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:146)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:117)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:81)
	at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build(SingleThreadedBuilder.java:51)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:128)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:309)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:194)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:107)
	at org.apache.maven.cli.MavenCli.execute(MavenCli.java:993)
	at org.apache.maven.cli.MavenCli.doMain(MavenCli.java:345)
	at org.apache.maven.cli.MavenCli.main(MavenCli.java:191)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced(Launcher.java:289)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:229)
	at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode(Launcher.java:415)
	at org.codehaus.plexus.classworlds.launcher.Launcher.main(Launcher.java:356)
Caused by: com.ibm.ws.repository.resolver.RepositoryResolutionException: org.osgi.service.resolver.ResolutionException: Unable to resolve LpmResource [resource=jsp-2.2]: missing requirement com.ibm.ws.repository.resolver.internal.resource.InstallableEntityRequirement [jsp-2.2] [caused by: Unable to resolve FeatureResource [symbolicName=com.ibm.websphere.appserver.jsp-2.2, version=1.0.0, location=repo]: missing requirement com.ibm.ws.repository.resolver.internal.resource.InstallableEntityRequirement [com.ibm.websphere.appserver.servlet-3.0]]
	at com.ibm.ws.repository.resolver.RepositoryResolver.resolve(RepositoryResolver.java:289)
	at com.ibm.ws.install.internal.InstallKernelMap.singleFileResolve(InstallKernelMap.java:503)
	... 27 more
Caused by: org.osgi.service.resolver.ResolutionException: Unable to resolve LpmResource [resource=jsp-2.2]: missing requirement com.ibm.ws.repository.resolver.internal.resource.InstallableEntityRequirement [jsp-2.2] [caused by: Unable to resolve FeatureResource [symbolicName=com.ibm.websphere.appserver.jsp-2.2, version=1.0.0, location=repo]: missing requirement com.ibm.ws.repository.resolver.internal.resource.InstallableEntityRequirement [com.ibm.websphere.appserver.servlet-3.0]]
	at org.apache.felix.resolver.ResolutionError.toException(ResolutionError.java:42)
	at org.apache.felix.resolver.ResolverImpl.doResolve(ResolverImpl.java:391)
	at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:377)
	at org.apache.felix.resolver.ResolverImpl.resolve(ResolverImpl.java:370)
	at com.ibm.ws.repository.resolver.RepositoryResolver.resolve(RepositoryResolver.java:282)
	... 28 more
```

Other features where this situation occurs are as follows:
```
com.ibm.websphere.appserver.jcaInboundSecurity-1.0
  depends on com.ibm.websphere.appserver.jca-1.6, tolerates 1.7

com.ibm.websphere.appserver.adminSecurity-1.0
com.ibm.websphere.appserver.appSecurity-1.0
com.ibm.websphere.appserver.jaspic-1.1
com.ibm.websphere.appserver.jsp-2.2
com.ibm.websphere.appserver.jwt-1.0
com.ibm.websphere.appserver.passwordUtilities-1.0
com.ibm.websphere.appserver.restConnector-2.0
com.ibm.websphere.appserver.restHandler-1.0
com.ibm.wsspi.appserver.webBundle-1.0
com.ibm.wsspi.appserver.webBundleSecurity-1.0
  depends on com.ibm.websphere.appserver.servlet-3.0, tolerates 3.1, 4.0
```

**Comment by tjwatson** [comment]
How are you encoding your feature capabilities and requirements?

I would have thought the following capability for the feature `com.ibm.websphere.appserver.servlet-3.0`:

`Provide-Capability: libert.feature; liberty.feature=com.ibm.websphere.appserver.servlet; version:Version=3.0`

And for the requirement depends on `com.ibm.websphere.appserver.servlet-3.0; ibm.tolerates="3.1, 4.0"`

`Require-Capability: libert.feature; filter:="(&(liberty.feature=com.ibm.websphere.appserver.servlet)(|(version=3.0)(version=3.1)(version=4.0)))"`

This way the requirement can match any of the following capabilities:
`libert.feature; liberty.feature=com.ibm.websphere.appserver.servlet; version:Version=3.0`
`libert.feature; liberty.feature=com.ibm.websphere.appserver.servlet; version:Version=3.1`
`libert.feature; liberty.feature=com.ibm.websphere.appserver.servlet; version:Version=4.0`

**Comment by tjwatson** [comment]
Define "correct version".  The feature stated that it can tolerate a specific list of versions.  Any of them should be valid to use.

**Comment by tjwatson** [comment]
The resolve context that I assume you implement should be returning an ordered list of capabilities, the resolver then is supposed to honor that.  The issue with using the OSGi resolver is that we have no good way to express the propagation of toleration over a dependency chain.  If you really want to determine a valid set of features I think you should be using the same feature resolver the runtime uses.

**Comment by tjwatson** [comment]
Sorry to keep piling on new comments ...

Another option may involve updating com.ibm.ws.repository.resolver.internal.RepositoryResolveContext.match(Requirement, Capability) such that it will match any of the tolerated feature capabilities.

**Comment by ericglau** [comment]
Fixed by #3854
